src/credentials/providers/sharedCredentialsProvider.ts

```diff
@@ -0,0 +1,192 @@
+        private readonly profileName: string,
+        private readonly allSharedCredentialProfiles: Map<string, Profile>
+    ) {
+        const profile = this.allSharedCredentialProfiles.get(profileName)
+
+        if (!profile) {
+            throw new Error(`Profile not found: ${profileName}`)
+        }
+
+        this.profile = profile
+    }
+
+    public getCredentialsProviderId(): string {
+        return makeCredentialsProviderId({
+            credentialType: SharedCredentialsProvider.getCredentialsType(),
+            credentialTypeId: this.profileName
+        })
+    }
+
+    public getHashCode(): number {
+        return getStringHash(JSON.stringify(this.profile))
+    }
+
+    public getDefaultRegion(): string | undefined {
+        return this.profile[SHARED_CREDENTIAL_PROPERTIES.REGION]
+    }
+
+    /**
+     * Returns undefined if the Profile is valid, else a string indicating what is invalid
+     */
+    public validate(): string | undefined {
+        const expectedProperties: string[] = []
+
+        if (this.hasProfileProperty(SHARED_CREDENTIAL_PROPERTIES.ROLE_ARN)) {
+            return this.validateSourceProfileChain()
+        } else if (this.hasProfileProperty(SHARED_CREDENTIAL_PROPERTIES.CREDENTIAL_PROCESS)) {
+            // No validation. Don't check anything else.
+
+            return undefined
+        } else if (this.hasProfileProperty(SHARED_CREDENTIAL_PROPERTIES.AWS_SESSION_TOKEN)) {
+            expectedProperties.push(
+                SHARED_CREDENTIAL_PROPERTIES.AWS_ACCESS_KEY_ID,
+                SHARED_CREDENTIAL_PROPERTIES.AWS_SECRET_ACCESS_KEY
+            )
+        } else if (this.hasProfileProperty(SHARED_CREDENTIAL_PROPERTIES.AWS_ACCESS_KEY_ID)) {
+            expectedProperties.push(SHARED_CREDENTIAL_PROPERTIES.AWS_SECRET_ACCESS_KEY)
+        }
+
+        const missingProperties = this.getMissingProperties(expectedProperties)
+        if (missingProperties.length !== 0) {
+            return `Profile ${this.profileName} is missing properties: ${missingProperties.join(', ')}`
+        }
+
+        return undefined
+    }
+
+    public async getCredentials(): Promise<AWS.Credentials> {
+        const validationMessage = this.validate()
+        if (validationMessage) {
+            throw new Error(`Profile ${this.profileName} is not a valid Credential Profile: ${validationMessage}`)
+        }
+
+        const provider = new AWS.CredentialProviderChain([this.makeCredentialsProvider()])
+
+        return provider.resolvePromise()
+    }
+
+    private hasProfileProperty(propertyName: string): boolean {
+        return !!this.profile[propertyName]
+    }
+
+    private getMissingProperties(propertyNames: string[]): string[] {
+        return propertyNames.filter(propertyName => !this.profile[propertyName])
+    }
+
+    /**
+     * Returns undefined if the Profile Chain is valid, else a string indicating what is invalid
+     */
+    private validateSourceProfileChain(): string | undefined {
```

Comment:
I'm curious, what happens if we don't have this safeguard in place? Does the JS SDK handle this? Is this necessary to handle on our end?

Reply:
This is necessary to handle on our end. I forget the behavior but its bad. It either fully kills VS Code, or fully locks up VS Code.

Even if the SDK handled it, it is important to provide useful information to users so that they can remedy the situation. In this case, the chain of profiles is logged out.